File: src-server/src/api/exchange_rates.rs
use std::sync::Arc;

use crate::{api::shared::trigger_full_portfolio_recalc, error::ApiResult, main_lib::AppState};
use axum::{
    extract::{Path, State},
    http::StatusCode,
    routing::{delete, get, put},
    Json, Router,
};
use wealthfolio_core::fx::{ExchangeRate, NewExchangeRate};

async fn get_latest_exchange_rates(
    State(state): State<Arc<AppState>>,
) -> ApiResult<Json<Vec<ExchangeRate>>> {
    let rates = state.fx_service.get_latest_exchange_rates()?;
    Ok(Json(rates))
}

async fn update_exchange_rate(
    State(state): State<Arc<AppState>>,
    Json(rate): Json<ExchangeRate>,
) -> ApiResult<Json<ExchangeRate>> {
    let updated = state
        .fx_service
        .update_exchange_rate(&rate.from_currency, &rate.to_currency, rate.rate)
        .await?;
    trigger_full_portfolio_recalc(state.clone());
    Ok(Json(updated))
}

async fn add_exchange_rate(
    State(state): State<Arc<AppState>>,
    Json(new_rate): Json<NewExchangeRate>,
) -> ApiResult<Json<ExchangeRate>> {
    let added = state.fx_service.add_exchange_rate(new_rate).await?;
    trigger_full_portfolio_recalc(state.clone());
    Ok(Json(added))
}

async fn delete_exchange_rate(
    Path(id): Path<String>,
    State(state): State<Arc<AppState>>,
) -> ApiResult<StatusCode> {
    state.fx_service.delete_exchange_rate(&id).await?;
    trigger_full_portfolio_recalc(state);
    Ok(StatusCode::NO_CONTENT)
}

pub fn router() -> Router<Arc<AppState>> {
    Router::new()
        .route("/exchange-rates/latest", get(get_latest_exchange_rates))
        .route(
            "/exchange-rates",
            put(update_exchange_rate).post(add_exchange_rate),
        )
        .route("/exchange-rates/{id}", delete(delete_exchange_rate))
}

File: src-front/adapters/shared/exchange-rates.ts
// Exchange Rates Commands
import type { ExchangeRate } from "@/lib/types";

import { invoke, logger } from "./platform";

export const getExchangeRates = async (): Promise<ExchangeRate[]> => {
  try {
    return await invoke<ExchangeRate[]>("get_latest_exchange_rates");
  } catch (err) {
    logger.error("Error fetching exchange rates.");
    throw err;
  }
};

export const updateExchangeRate = async (updatedRate: ExchangeRate): Promise<ExchangeRate> => {
  return invoke<ExchangeRate>("update_exchange_rate", { rate: updatedRate });
};

export const addExchangeRate = async (newRate: Omit<ExchangeRate, "id">): Promise<ExchangeRate> => {
  return invoke<ExchangeRate>("add_exchange_rate", { newRate });
};

export const deleteExchangeRate = async (rateId: string): Promise<void> => {
  return invoke<void>("delete_exchange_rate", { rateId });
};
